dali/pipeline/data/types.cc

```diff
@@ -34,36 +34,106 @@ const auto &_type_info_##Id = TypeTable::GetTypeID<Type>()
 
 namespace dali {
 
-TypeTable &TypeTable::instance() {
-  static TypeTable singleton;
+namespace detail {
+
+class ScatterGatherPool {
+ public:
+  static void Copy(void* dst, const void** srcs, const Index* sizes, int n, int element_size, cudaStream_t stream);
+
+ private:
+  static ScatterGatherPool& instance();
+
+  // ScatterGatherPool should only be referenced through its static members
+  ScatterGatherPool() {}
+
+  spinlock lock_;
+  std::unordered_map<cudaStream_t, std::unique_ptr<kernels::ScatterGatherGPU>>
```

Comment:
I don't think this extra level of indirection is necessary.
```suggestion
  std::unordered_map<cudaStream_t, kernels::ScatterGatherGPU>
```